node/node_test.go

```diff
@@ -622,6 +623,50 @@ func TestManualConditionsPreserved(t *testing.T) {
 
 	t.Log(newNode.Status.Conditions)
 }
+
+func TestNodePingSingleInflight(t *testing.T) {
+	const pingTimeout = 100 * time.Millisecond
+	c := testclient.NewSimpleClientset()
+	testP := &testNodeProviderPing{}
+
+	var calls, finished int64
+	testP.customPingFunction = func(context.Context) error {
+		atomic.AddInt64(&calls, 1)
+		time.Sleep(2 * time.Second)
+		atomic.AddInt64(&finished, 1)
+		return nil
+	}
+
+	nodes := c.CoreV1().Nodes()
+
+	testNode := testNode(t)
+
+	node, err := NewNodeController(testP, testNode, nodes, WithNodePingInterval(10*time.Millisecond), WithNodePingTimeout(pingTimeout))
+	assert.NilError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	start := time.Now()
+	go node.nodePingController.run(ctx)
+	firstPing, err := node.nodePingController.getResult(ctx)
+	assert.NilError(t, err)
+	timeTakenToCompleteFirstPing := time.Since(start)
+	assert.Assert(t, timeTakenToCompleteFirstPing < pingTimeout*5, "Time taken to complete first ping: %v", timeTakenToCompleteFirstPing)
+
+	assert.Assert(t, cmp.Error(firstPing.error, context.DeadlineExceeded.Error()))
+	assert.Assert(t, is.Equal(int64(1), atomic.LoadInt64(&calls)))
+	assert.Assert(t, is.Equal(int64(0), atomic.LoadInt64(&finished)))
+
+	// Wait until the first sleep finishes
+	for atomic.LoadInt64(&finished) == 0 {
```

Comment:
Just a test, but this is going to cause a tight loop and eat up CPU.
Should be better on go1.14, but really problematic <=go1.13.

Maybe a cond variable or a a channel to synchronize this.

Reply:
We have waitable int, in mock_test.go. Mind if I just use that?

Reply:
As long as it works I'm happy 😸